src/apps/metastreet/ethereum/metastreet.lending-v2.contract-position-fetcher.ts

```diff
@@ -0,0 +1,362 @@
+    const limit: number = BigNumber.from(_params.definition.limit).lt(1e15)
+      ? 0
+      : BigNumber.from(_params.definition.limit).div(1e15).toNumber() / 1000;
+    const apr: BigNumber = BigNumber.from(_params.definition.rate).mul(365 * 86400);
+    const rate: number = Math.round(apr.lt(1e15) ? 0 : apr.div(1e15).toNumber() / 10);
+    const labelPrefix =
+      collateralTokenName && currencyTokenSymbol ? `${collateralTokenName} / ${currencyTokenSymbol} - ` : '';
+
+    // e.g. "Wrapped Cryptopunks / DAI - 30 Day, 10%, 630000 DAI"
+    return `${labelPrefix}${duration} Day, ${rate}%, ${limit} ${currencyTokenSymbol}`;
+  }
+
+  async getPositionsForBalances() {
+    return this.appToolkit.getAppContractPositions<DataProps>({
+      appId: this.appId,
+      network: this.network,
+      groupIds: [this.groupId],
+    });
+  }
+
+  async getBalances(_address: string): Promise<ContractPositionBalance<DataProps>[]> {
+    const multicall = this.appToolkit.getMulticall(this.network);
+    const address = await this.getAccountAddress(_address);
+    if (address === ZERO_ADDRESS) return [];
+
+    /* Get ticks that the account has interacted with */
+    const data = await gqlFetch<GetAllDepositsResponse>({
+      endpoint: SUBGRAPH_URL,
+      query: getAllDepositsQuery(address),
+    });
+
+    /* Tick IDs */
+    const ticks = new Set();
+
+    /* Add active positions to the set of tick IDs */
+    data.pools.forEach(p => {
+      p.deposits.forEach(d => {
+        /* Only add a position if user shares is non-zero */
+        if (
+          !BigNumber.from(d.shares).eq(constants.Zero) ||
+          d.redemptions.some(r => !BigNumber.from(r.shares).eq(constants.Zero))
+        ) {
+          ticks.add(d.tick.id);
+        }
+      });
+    });
+
+    /* Get all positions from all pools */
+    const contractPositions = await this.getPositionsForBalances();
+
+    /* Filter out positions based on account's active positions */
+    const filteredPositions = contractPositions.filter(cp => ticks.has(cp.dataProps.positionKey));
+
+    const balances = await Promise.all(
+      filteredPositions.map(async contractPosition => {
+        const contract = multicall.wrap(this.getContract(contractPosition.address));
+        const balancesRaw = await this.getTokenBalancesPerPosition({ address, contract, contractPosition, multicall });
+        const allTokens = contractPosition.tokens.map((cp, idx) =>
+          drillBalance(cp, balancesRaw[idx]?.toString() ?? '0', { isDebt: cp.metaType === MetaType.BORROWED }),
+        );
+
+        const tokens = allTokens.filter(v => Math.abs(v.balanceUSD) > 0.01);
+        const balanceUSD = sumBy(tokens, t => t.balanceUSD);
+
+        const balance = { ...contractPosition, tokens, balanceUSD };
+        return balance;
+      }),
+    );
+    return balances;
+  }
+
+  async getTokenBalancesPerPosition(_params: GetTokenBalancesParams<Pool, DataProps>): Promise<BigNumberish[]> {
```

Comment:
tip: 
You can deconstruct params in the function signature
`async getTokenBalancesPerPosition(_params: GetTokenBalancesParams<Pool, DataProps>):`
vs 
`async getTokenBalancesPerPosition( {contract, contractPosition, address}: GetTokenBalancesParams<Pool, DataProps>):`